c2corg_ui/static/js/documentediting.js

```diff
@@ -0,0 +1,269 @@
+goog.provide('app.DocumentEditingController');
+goog.provide('app.documentEditingDirective');
+
+goog.require('app');
+goog.require('ol.format.GeoJSON');
+goog.require('ol.geom.Point');
+
+
+
+/**
+ * @constructor
+ * @return {angular.Directive} The directive specs.
+ * @ngInject
+ */
+app.documentEditingDirective = function() {
+  return {
+    restrict: 'A',
+    scope: true,
+    controller: 'appDocumentEditingController',
+    controllerAs: 'editCtrl',
+    bindToController: true
+  };
+};
+
+
+app.module.directive('appDocumentEditing', app.documentEditingDirective);
+
+
+
+/**
+ * @param {angular.Scope} $scope Scope.
+ * @param {angular.JQLite} $element Element.
+ * @param {angular.Attributes} $attrs Attributes.
+ * @param {angular.$http} $http
+ * @param {string} apiUrl Base URL of the API.
+ * @constructor
+ * @ngInject
+ * @export
+ */
+app.DocumentEditingController = function($scope, $element, $attrs, $http,
+    apiUrl) {
+
+  /**
+   * @type {angular.Scope}
+   * @private
+   */
+  this.scope_ = $scope;
+
+  /**
+   * @type {angular.$http}
+   * @private
+   */
+  this.http_ = $http;
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.apiUrl_ = apiUrl;
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.module_ = $attrs['appDocumentEditing'];
+  goog.asserts.assert(goog.isDef(this.module_));
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.modelName_ = $attrs['appDocumentEditingModel'];
+
+  /**
+   * @type {?number}
+   * @private
+   */
+  this.id_ = null;
+
+  /**
+   * @type {?string}
+   * @private
+   */
+  this.culture_ = null;
+
+  if ('appDocumentEditingId' in $attrs &&
+      'appDocumentEditingCulture' in $attrs) {
+    this.id_ = $attrs['appDocumentEditingId'];
+    this.culture_ = $attrs['appDocumentEditingCulture'];
+    // Get document attributes from the API to feed the model:
+    this.http_.get(this.buildUrl_('read')).then(
+        goog.bind(this.successRead_, this),
+        goog.bind(this.errorRead_, this)
+    );
+  }
+};
+
+
+/**
+ * @const
+ * @type {string}
+ */
+app.DocumentEditingController.VIEW_PROJ = 'EPSG:4326';
+
+
+/**
+ * @const
+ * @type {string}
+ */
+app.DocumentEditingController.DATA_PROJ = 'EPSG:3857';
+
+
+/**
+ * @param {string} type Type of URL.
+ * @return {string} URL.
+ * @private
+ */
+app.DocumentEditingController.prototype.buildUrl_ = function(type) {
+  var url = this.apiUrl_;
+  url += url.substr(-1) != '/' ? '/' : '';
+  url += this.module_;
+  switch (type) {
+    case 'read':
+      url += '/' + this.id_ + '?l=' + this.culture_;
```

Comment:
A nice way to construct an URL is also do something like the following:

      this.url_ =
          'http://provelobern-geomapfish.prod.sig.cloud.camptocamp.net/' +
          '{profile}/viaroute?loc={from}&loc={to}&instructions=false&alt=false' +
          '&z={zoom}&output=json';

     var url = this.osrmUrl_
        .replace('{profile}', this.getProfile_())
        .replace('{from}', from[1] + ',' + from[0])
        .replace('{to}', to[1] + ',' + to[0])
        .replace('{zoom}', '20');